internal/repository/master_index.go

```diff
@@ -237,6 +237,27 @@ func (mi *MasterIndex) Each(ctx context.Context) <-chan restic.PackedBlob {
 	return ch
 }
 
+func (mi *MasterIndex) MergeFinalIndexes() {
+	mi.idxMutex.Lock()
+	defer mi.idxMutex.Unlock()
+
+	var firstFinalIndex *Index
+	newIdx := mi.idx[:0]
+
+	for _, idx := range mi.idx {
```

Comment:
I think the loop can be avoided.

Say MasterIndex where defined as

```go
type MasterIndex struct {
	idx          map[*Index]struct{} // non-final indexes
	frozenIdx    *Index
	pendingBlobs restic.BlobSet
	idxMutex     sync.RWMutex
}
```

then to merge an index by id, first look it up in the map by pointer and delete it if present, then merge it. I suppose LoadIndex could bypass the map entirely.

The Index.final field could perhaps be removed, if the code paths touching the final/frozen index and the other ones are well-separated.

Reply:
That way, there's also one Index that is usually the largest, which is a good candidate for starting lookups.

We may have to worry about contention: eventually all ids end up in one Index, and an Index uses a Mutex even for lookups. But switching to RWMutex or even sync.Map should be easy enough.

Reply:
I also thought about directly merging from LoadIndex. However, there is a problem: When correctly using the "superseded" field in the index file, you should remove the superseded indexes - if some are present - from MasterIndex. As you cannot say if a later-loaded index will supersed the actual index, we shouldn't load it. However, if it is already merged, we are unable to remove the entries after.

I added this correct superseded logic to #2718, but maybe I should make a separate PR for that topic...

Reply:
Too bad. I hadn't thought of that.